inference-engine/tests/functional/plugin/shared/src/single_layer_tests/gather_nd.cpp

```diff
@@ -0,0 +1,105 @@
+#include <vector>
+
+#include "common_test_utils/common_utils.hpp"
+#include "functional_test_utils/precision_utils.hpp"
+#include "functional_test_utils/skip_tests_config.hpp"
+
+#include "single_layer_tests/gather_nd.hpp"
+
+namespace IE = InferenceEngine;
+
+namespace LayerTestsDefinitions {
+
+std::string GatherNDLayerTest::getTestCaseName(const testing::TestParamInfo<gatherNDParams> &obj) {
+    gatherNDLayerParams params;
+    IE::Precision netPrc, indicesPrc;
+    IE::Layout inLayout, outLayout;
+    LayerTestsUtils::TargetDevice targetDevice;
+    std::map<std::string, std::string> config;
+
+    std::tie(params, netPrc, indicesPrc, inLayout, outLayout, targetDevice, config) = obj.param;
+
+    std::size_t batchDims;
+    IE::SizeVector inputShape, indicesShape;
+    std::tie(batchDims, inputShape, indicesShape) = params;
+
+    std::ostringstream result;
+    result << "IS=" << CommonTestUtils::vec2str(inputShape) << "_";
+    result << "indicesShape=" << CommonTestUtils::vec2str(indicesShape) << "_";
+    result << "batchDims=" << batchDims << "_";
+    result << "inPRC=" << netPrc.name() << "_";
+    result << "indicesPRC=" << indicesPrc.name() << "_";
+    result << "outPRC=" << netPrc.name() << "_";
+    result << "inL=" << inLayout << "_";
+    result << "outL=" << outLayout << "_";
+    result << "trgDev=" << targetDevice << "_";
+    return result.str();
+}
+
+void GatherNDLayerTest::SetUp() {
+    gatherNDLayerParams layerParams;
+    IE::Precision netPrc, indicesPrc;
+    std::map<std::string, std::string> config;
+
+    std::tie(layerParams, netPrc, indicesPrc, inLayout, outLayout, targetDevice, config) =
+        this->GetParam();
+
+    // FIXME Those precisions are used as actual network/layer data types
+    // Can't change it to be FP16
+    inPrc = InferenceEngine::Precision::FP32;
+    outPrc = InferenceEngine::Precision::FP32;
+
+    configuration.insert(begin(config), end(config));
+
+    // Unpack layer params
+    std::size_t batchDims;
+    IE::SizeVector inputShape, indicesShape;
+    std::tie(batchDims, inputShape, indicesShape) = layerParams;
+
+    auto gpInPrc = FuncTestUtils::PrecisionUtils::convertIE2nGraphPrc(netPrc);
+    auto params = ngraph::builder::makeParams(gpInPrc, {inputShape});
+    auto paramsOuts = ngraph::helpers::convert2OutputVector(
+        ngraph::helpers::castOps2Nodes<ngraph::op::Parameter>(params));
+
+    // Generate indices values
+    const auto indices = [&] {
+        const auto indicesCount = std::accumulate(
+            begin(indicesShape), prev(end(indicesShape)), 1ull, std::multiplies<std::size_t>{});
+        const auto indexSize = indicesShape.back();
+
+        const auto maxDim = *std::max_element(begin(inputShape), end(inputShape));
+        std::random_device rd;
+        std::mt19937 gen(rd());
+        std::uniform_int_distribution<int> distrib(1, maxDim);
+
+        std::vector<int> indicesValues;
+        indicesValues.reserve(indicesCount * indexSize);
```

Comment:
How about a code like this?

```
std::vector<int> indicesValues(indicesCount * indexSize);
std::generate (indicesValues.begin(), indicesValues.end(), Generator);
```

Reply:
Indexes depend  on current position in the tensor.  
If indexSize == 3, indices vector looks like [c, h, w, c, h, w, ...], where c < C, h < H, w < W for tensor with CxHxW size.

Reply:
ok, I see